mkosi/vmspawn.py

```diff
@@ -0,0 +1,165 @@
+        if config.ephemeral:
+            fname = stack.enter_context(copy_ephemeral(config, config.output_dir_or_cwd() / config.output))
+        else:
+            fname = config.output_dir_or_cwd() / config.output
+
+        # Make sure vmspawn can access the ephemeral copy. Not required for directory output because we don't pass that
+        # directly to vmspawn, but indirectly via virtiofsd.
+        if config.output_format != OutputFormat.directory:
+            os.chown(fname, INVOKING_USER.uid, INVOKING_USER.gid)
+
+        if config.output_format == OutputFormat.disk and config.runtime_size:
+            run(["systemd-repart",
+                 "--definitions", "",
+                 "--no-pager",
+                 f"--size={config.runtime_size}",
+                 "--pretty=no",
+                 "--offline=yes",
+                 fname])
+
+        if kernel and (KernelType.identify(kernel) != KernelType.uki or not config.architecture.supports_smbios()):
+            kcl = config.kernel_command_line + config.kernel_command_line_extra
+        else:
+            kcl = config.kernel_command_line_extra
+
+        for tree in config.runtime_trees:
+            target = Path('/root/src') / (tree.target or tree.source.name)
+            cmdline += ["--bind", f"{tree.source}:{target}"]
+
+        root = None
+        if kernel:
+            cmdline += ["--kernel", kernel]
+
+            if any(s.startswith("root=") for s in kcl):
+                pass
+            elif config.output_format == OutputFormat.disk:
+                # We can't rely on gpt-auto-generator when direct kernel booting so synthesize a root=
+                # kernel argument instead.
+                root = finalize_root(find_partitions(fname))
+                if not root:
+                    die("Cannot perform a direct kernel boot without a root or usr partition")
+
+                kcl += [root]
+
+        if config.output_format == OutputFormat.directory:
+            cmdline += [f"--directory={fname}"]
+            if os.getuid() != 0:
+                cmdline += [f"--private-users={read_subrange(Path('/etc/subuid'))}:65536"]
+        else:
+            cmdline += [f"--image={fname}"]
+
+        cmdline += [
+            *args.cmdline,
+            "--",
+            *kcl,
+        ]
+
+        with spawn(
+            cmdline,
+            # On Debian/Ubuntu, only users in the kvm group can access /dev/kvm. The invoking user might be part of the
+            # kvm group, but the user namespace fake root user will definitely not be. Thus, we have to run qemu as the
+            # invoking user to make sure we can access /dev/kvm. Of course, if we were invoked as root, none of this
+            # matters as the root user will always be able to access /dev/kvm.
+            user=INVOKING_USER.uid if not INVOKING_USER.invoked_as_root else None,
+            group=INVOKING_USER.gid if not INVOKING_USER.invoked_as_root else None,
```

Comment:
These two shouldn't be necessary either, we'll just run nspawn as the current user, which will either be a regular user or root.

Reply:
Fixed in dc05caea.